compiler/src/vm/heap/object.rs

```diff
@@ -0,0 +1,257 @@
+use super::{pointer::Pointer, Heap};
+use crate::{
+    builtin_functions::BuiltinFunction,
+    compiler::{
+        hir_to_lir::HirToLir,
+        lir::{Instruction, Lir},
+    },
+    database::Database,
+    module::Module,
+};
+use itertools::Itertools;
+use num_bigint::BigInt;
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+#[derive(Clone)]
+pub struct Object {
+    pub reference_count: usize,
+    pub data: Data,
+}
+#[derive(Clone)]
+pub enum Data {
+    Int(Int),
+    Text(Text),
+    Symbol(Symbol),
+    Struct(Struct),
+    Closure(Closure),
+    Builtin(Builtin),
+}
+
+#[derive(Clone)]
+pub struct Int {
+    pub value: BigInt,
+}
+
+#[derive(Clone)]
+pub struct Text {
+    pub value: String,
+}
+
+#[derive(Clone)]
+pub struct Symbol {
+    // TODO: For better memory efficiency, store the symbol as a null-terminated
+    // string.
+    pub value: String,
+}
+
+#[derive(Default, Clone)]
+pub struct Struct {
+    // TODO: Choose a more efficient representation.
+    pub fields: Vec<(u64, Pointer, Pointer)>, // hash, key, and value
+}
+
+#[derive(Clone)]
+pub struct Closure {
+    pub captured: Vec<Pointer>,
+    pub num_args: usize,
+    pub body: Vec<Instruction>,
+}
+
+#[derive(Clone)]
+pub struct Builtin {
+    pub function: BuiltinFunction,
+}
+
+impl Struct {
+    pub fn from_fields(heap: &Heap, fields: HashMap<Pointer, Pointer>) -> Self {
+        let mut s = Self::default();
+        for (key, value) in fields {
+            s.insert(heap, key, value);
+        }
+        s
+    }
+    fn insert(&mut self, heap: &Heap, key: Pointer, value: Pointer) {
+        let hash = key.hash(heap);
+        for (existing_hash, existing_key, existing_value) in &mut self.fields {
+            if hash != *existing_hash {
+                continue;
+            }
+            if existing_key.equals(heap, key) {
+                *existing_value = value;
+                return;
+            }
+        }
+        let entry = (hash, key, value);
+        if let Some(index) = self
+            .fields
+            .iter()
+            .position(|(the_hash, _, _)| *the_hash > hash)
```

Comment:
If the fields are sorted, we could (should?) use binary search. Also, a single traversal is sufficient